File: Python/stats.py
"""
Python (Object-Oriented): Mean, Median, Mode for a list of integers.
- Reads one line of space-separated integers from stdin.
- Encapsulates statistical logic inside a class.
- Prints "None" if no value repeats (no mode).
"""

class StatisticsCalculator:
    def __init__(self, data):
        # Store the list of integers
        self.data = data

    def mean(self):
        return sum(self.data) / len(self.data)

    def median(self):
        sorted_data = sorted(self.data)
        n = len(sorted_data)

        if n % 2 == 1:
            return float(sorted_data[n // 2])
        else:
            a = sorted_data[n // 2 - 1]
            b = sorted_data[n // 2]
            return (a + b) / 2.0

    def mode(self):
        # Count frequencies using a dictionary
        freq = {}
        for value in self.data:
            freq[value] = freq.get(value, 0) + 1

        max_freq = max(freq.values())

        # If all values occur once, there is no mode
        if max_freq == 1:
            return []

        # Return all values with maximum frequency (sorted for consistency)
        return sorted([k for k, v in freq.items() if v == max_freq])


def parse_input():
    line = input("Enter integers separated by spaces:\n").strip()
    if not line:
        return None

    tokens = line.split()
    try:
        return [int(token) for token in tokens]
    except ValueError:
        return None


if __name__ == "__main__":
    numbers = parse_input()

    if not numbers:
        print("Error: please enter at least one integer.")
    else:
        calculator = StatisticsCalculator(numbers)

        mu = calculator.mean()
        med = calculator.median()
        modes = calculator.mode()

        print(f"Count: {len(numbers)}")
        print(f"Mean: {mu:.2f}")

        # Print median as integer if whole number
        if med.is_integer():
            print(f"Median: {int(med)}")
        else:
            print(f"Median: {med:.2f}")

        if not modes:
            print("Mode: None")
        else:
            print("Mode:", *modes)

File: C/stats.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/*
  C (Procedural): Mean, Median, Mode for a list of integers.
  - Reads one line of space-separated integers from stdin.
  - Sorts the data.
  - Computes mean, median, and mode.
*/

#define MAX_LINE 4096

/* Comparator used by qsort (ascending order) */
static int cmp_ints(const void *a, const void *b)
{
    int x = *(const int *)a;
    int y = *(const int *)b;
    return (x > y) - (x < y);
}

/* Mean: average of all values */
static double mean(const int *a, int n)
{
    long long sum = 0;
    for (int i = 0; i < n; i++)
        sum += a[i];
    return (double)sum / (double)n;
}

/* Median computed after sorting the array */
static double median_after_sort(const int *a, int n)
{
    if (n % 2 == 1)
        return (double)a[n / 2];
    return ((double)a[n / 2 - 1] + (double)a[n / 2]) / 2.0;
}

int main(void)
{
    char line[MAX_LINE];

    printf("Enter integers separated by spaces:\n");
    if (!fgets(line, sizeof(line), stdin))
    {
        printf("Error: please enter at least one integer.\n");
        return 1;
    }

    /*
      Parsing:
      - Split the input line by whitespace (strtok).
      - Convert each token to an integer (strtol) and reject invalid tokens.
      - Store integers in a dynamically grown array (malloc/realloc).
    */
    int capacity = 16;
    int n = 0;
    int *a = (int *)malloc(sizeof(int) * capacity);
    if (!a)
    {
        fprintf(stderr, "Error: memory allocation failed.\n");
        return 1;
    }

    char *tok = strtok(line, " \t\r\n");
    while (tok != NULL)
    {
        char *end = NULL;
        long v = strtol(tok, &end, 10);

        if (end == tok || *end != '\0')
        {
            printf("Error: please enter at least one integer.\n");
            free(a);
            return 1;
        }

        if (n == capacity)
        {
            capacity *= 2;
            int *tmp = (int *)realloc(a, sizeof(int) * capacity);
            if (!tmp)
            {
                fprintf(stderr, "Error: memory allocation failed.\n");
                free(a);
                return 1;
            }
            a = tmp;
        }

        a[n++] = (int)v;
        tok = strtok(NULL, " \t\r\n");
    }

    if (n == 0)
    {
        printf("Error: please enter at least one integer.\n");
        free(a);
        return 1;
    }

    /* Sorting creates the ordering required for median and mode counting */
    qsort(a, n, sizeof(int), cmp_ints);

    double mu = mean(a, n);
    double med = median_after_sort(a, n);

    /*
      Mode:
      - After sorting, equal values are adjacent.
      - First pass finds the largest run length (max frequency).
      - If max frequency is 1, no value repeats, so the dataset has no mode.
      - Second pass prints all values whose run length equals max frequency.
    */
    int max_freq = 1;
    int curr_freq = 1;

    for (int i = 1; i < n; i++)
    {
        if (a[i] == a[i - 1])
        {
            curr_freq++;
        }
        else
        {
            if (curr_freq > max_freq)
                max_freq = curr_freq;
            curr_freq = 1;
        }
    }
    if (curr_freq > max_freq)
        max_freq = curr_freq;

    printf("Count: %d\n", n);
    printf("Mean: %.2f\n", mu);

    if (med == (double)((long long)med))
    {
        printf("Median: %lld\n", (long long)med);
    }
    else
    {
        printf("Median: %.2f\n", med);
    }

    if (max_freq == 1)
    {
        printf("Mode: None\n");
        free(a);
        return 0;
    }

    printf("Mode:");
    curr_freq = 1;
    for (int i = 1; i <= n; i++)
    {
        if (i < n && a[i] == a[i - 1])
        {
            curr_freq++;
        }
        else
        {
            if (curr_freq == max_freq)
            {
                printf(" %d", a[i - 1]);
            }
            curr_freq = 1;
        }
    }
    printf("\n");

    free(a);
    return 0;
}
